src/psyclone/psyir/transformations/replace_induction_variables.py

```diff
@@ -0,0 +1,272 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Module providing a transformation that removed induction variables from
+a loop. '''
+
+from __future__ import absolute_import
+
+from psyclone.core import AccessType, VariablesAccessInfo
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import (ArrayReference, Assignment, BinaryOperation,
+                                  Loop, Reference)
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+
+
+class ReplaceInductionVariables(Transformation):
+    '''Move all supported induction variables out of the loop, and replace
+    their usage inside the loop. For example:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.psyir.transformations import ReplaceInductionVariables
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> psyir = FortranReader().psyir_from_source("""
+    ... subroutine sub()
+    ...     integer :: i, im, ic, tmp(100)
+    ...     do i=1, 100
+    ...         im = i - 1
+    ...         ic = 2
+    ...         tmp(i) = ic * im
+    ...     enddo
+    ... end subroutine sub""")
+    >>> loop = psyir.walk(Loop)[0]
+    >>> ReplaceInductionVariables().apply(loop)
+    >>> print(FortranWriter()(psyir))
+
+    will generate:
+
+    .. code-block:: fortran
+
+        subroutine sub()
+          integer :: i
+          integer :: im
+          integer :: ic
+          integer, dimension(100) :: tmp
+
+          do i = 1, 100, 1
+            tmp(i) = 2 * (i - 1)
+          enddo
+          ic = 2
+          im = i - 1 - 1
+
+        end subroutine sub
+
+    After the loop the replaced assignments to the induction variables are
+    added so these variables will have the correct value if they should be
+    used elsewhere.
+
+    The following restrictions apply for the assignment to an induction
+    variable:
+
+    * the variable must be a non-structure scalar.
+    * none of variables on the right-hand side can be written in the loop body
+      (the loop variable is written in the Loop statement, not in
+      the body, so it can be used).
+    * the assigned variable must not be read before the assignment.
+    * the assigned variable cannot occur on the right-hand side
+      (e.g. `k = k + 3`).
+    * there must only be one assignment to the variable.
+
+    '''
+    def __str__(self):
+        return "Replaces all induction variables in a loop."
+
+    # ------------------------------------------------------------------------
+    @staticmethod
+    def _replace_references(psyir, original, replacement):
```

Comment:
This could possibly be a generic node functionality, since the walk(Reference) could also be walk(Node) as the equals already check the types match. For instance:
`node.replace_all(original, replacement)`
or
`node.replace_subexpressions(original, replacement)`
but maybe this is for another PR.

Reply:
Agreed with both points: could be useful, and happy to move this if we have the same requirement elsewhere, for now I just leave it here.